File: packages/sdk-java/src/main/java/io/settler/clients/CurrencyClient.java
package io.settler.clients;

import com.fasterxml.jackson.databind.JsonNode;
import io.settler.exceptions.*;
import io.settler.http.HttpExecutor;

import java.util.*;

/**
 * Client for multi-currency and FX operations.
 */
public class CurrencyClient {
    private static final String BASE_PATH = "/currency";

    private final HttpExecutor http;

    public CurrencyClient(HttpExecutor http) {
        this.http = http;
    }

    /**
     * Convert an amount to a target currency.
     *
     * @param value        the amount to convert
     * @param fromCurrency source currency code (e.g., "USD")
     * @param toCurrency   target currency code (e.g., "EUR")
     * @param date         optional date for historical rates (ISO 8601)
     * @return the conversion result
     * @throws SettlerException if the request fails
     */
    public JsonNode convert(double value, String fromCurrency, String toCurrency, String date) throws SettlerException {
        Map<String, Object> body = new LinkedHashMap<>();
        Map<String, Object> amount = new LinkedHashMap<>();
        amount.put("value", value);
        amount.put("currency", fromCurrency);
        body.put("amount", amount);
        body.put("toCurrency", toCurrency);
        if (date != null) {
            body.put("date", date);
        }
        return http.post(BASE_PATH + "/convert", body);
    }

    /**
     * Get the FX rate for a currency pair.
     *
     * @param fromCurrency source currency code
     * @param toCurrency   target currency code
     * @param date         optional date for historical rates (ISO 8601)
     * @return the FX rate data
     * @throws SettlerException if the request fails
     */
    public JsonNode getFxRate(String fromCurrency, String toCurrency, String date) throws SettlerException {
        Map<String, String> params = new LinkedHashMap<>();
        params.put("fromCurrency", fromCurrency);
        params.put("toCurrency", toCurrency);
        if (date != null) {
            params.put("date", date);
        }
        return http.get(BASE_PATH + "/fx-rate", params);
    }
}

File: packages/sdk-csharp/Clients/CurrencyClient.cs
using Settler.Sdk.Models;

namespace Settler.Sdk;

/// <summary>Client for currency and FX operations.</summary>
public sealed class CurrencyClient
{
    private readonly SettlerClient _client;

    internal CurrencyClient(SettlerClient client) => _client = client;

    /// <summary>Convert an amount to a target currency.</summary>
    public async Task<DataResponse<ConversionResult>> ConvertAsync(
        decimal value, string fromCurrency, string toCurrency, string? date = null,
        CancellationToken cancellationToken = default)
    {
        var body = new Dictionary<string, object>
        {
            ["amount"] = new { value, currency = fromCurrency },
            ["toCurrency"] = toCurrency,
        };
        if (date != null) body["date"] = date;

        return await _client.RequestAsync<DataResponse<ConversionResult>>(
            HttpMethod.Post, "currency/convert", body: body, cancellationToken: cancellationToken);
    }

    /// <summary>Get the FX rate for a currency pair.</summary>
    public async Task<DataResponse<FxRateResult>> GetFxRateAsync(
        string fromCurrency, string toCurrency, string? date = null,
        CancellationToken cancellationToken = default)
    {
        var query = new Dictionary<string, string>
        {
            ["fromCurrency"] = fromCurrency,
            ["toCurrency"] = toCurrency,
        };
        if (date != null) query["date"] = date;

        return await _client.RequestAsync<DataResponse<FxRateResult>>(
            HttpMethod.Get, "currency/fx-rate", query: query, cancellationToken: cancellationToken);
    }
}
